lib/bye_bye_bye/utils.ex

```diff
@@ -0,0 +1,173 @@
+    time = DateTime.to_time(datetime)
+
+    if Time.before?(time, ~T[03:00:00]) do
+      datetime
+      |> DateTime.add(-1, :day)
+      |> DateTime.to_date()
+    else
+      DateTime.to_date(datetime)
+    end
+  end
+
+  def build_cancellation_entity({trip_id, schedule}, now) do
+    route_id = schedule |> List.first() |> get_in(["relationships", "route", "data", "id"])
+
+    %FeedEntity{
+      id: trip_id,
+      trip_update: %TripUpdate{
+        timestamp: now,
+        trip: %TripDescriptor{
+          trip_id: trip_id,
+          route_id: route_id,
+          schedule_relationship: "CANCELED"
+        },
+        stop_time_update: Enum.map(schedule, &schedule_to_stop_time_update/1)
+      }
+    }
+  end
+
+  defp schedule_to_stop_time_update(schedule) do
+    %StopTimeUpdate{
+      stop_id: schedule["relationships"]["stop"]["data"]["id"],
+      stop_sequence: schedule["attributes"]["stop_sequence"],
+      schedule_relationship: "SKIPPED"
+    }
+  end
+
+  @doc """
+  Gets schedules affected by a cancellation alert.
+  Returns a map of trip_id to schedule entries.
+
+  ## Parameters
+    * `alert` - The alert to process
+    * `now` - Current time as DateTime, used to determine service day
+  """
+  def get_affected_schedules(alert, now) do
+    period_params =
+      alert["attributes"]["active_period"]
+      |> Enum.map(&period_params(&1, now))
+      |> Enum.filter(& &1)
+
+    trips =
+      alert["attributes"]["informed_entity"]
+      |> Enum.filter(fn entity -> !entity["stop"] && entity["trip"] end)
+      |> Enum.map(& &1["trip"])
+      |> then(fn trips ->
+        case trips do
+          [] -> []
+          _ -> [%{trips: trips}]
+        end
+      end)
+
+    routes =
+      alert["attributes"]["informed_entity"]
+      |> Enum.filter(fn entity -> !entity["stop"] && !entity["trip"] && entity["route"] end)
+      |> Enum.map(& &1["route"])
+      |> then(fn routes ->
+        case routes do
+          [] -> []
+          _ -> [%{routes: routes}]
+        end
+      end)
+
+    params = trips ++ routes
+
+    params
+    |> Enum.flat_map(&Enum.map(period_params, fn params -> Map.merge(params, &1) end))
+    |> Enum.map(&MbtaClient.get_schedules/1)
+    |> Enum.map(fn {:ok, schedules} -> schedules end)
+    |> Enum.reduce(%{}, &Map.merge(&1, &2))
+  end
+
+  defp service_day_times(now) do
+    now = DateTime.shift_zone!(now, "America/New_York")
+    time = DateTime.to_time(now)
+
+    if Time.before?(time, ~T[03:00:00]) do
+      {
+        now
+        |> DateTime.add(-1, :day)
+        |> DateTime.to_date()
+        |> DateTime.new!(~T[03:00:00], "America/New_York"),
+        now |> DateTime.to_date() |> DateTime.new!(~T[02:59:59], "America/New_York")
+      }
+    else
+      {
+        DateTime.to_date(now) |> DateTime.new!(~T[03:00:00], "America/New_York"),
+        DateTime.add(now, 1, :day)
+        |> DateTime.to_date()
+        |> DateTime.new!(~T[02:59:59], "America/New_York")
+      }
+    end
+  end
```

Comment:
suggestion:
I think you could simplify this by taking advantage of your existing `service_day/1`:
```suggestion
  defp service_day_times(now) do
    service_day = service_day(now)
    start_dt = DateTime.new(service_day, ~T[03:00:00], "America/New_York")

    end_dt =
      service_day
      |> DateTime.add(1, :day)
      |> DateTime.new(~T[03:00:00], "America/New_York")

    {start_dt, end_dt}
  end
```

Reply:
Good call.